src/tools/isabelle_executor.py

```diff
@@ -437,26 +453,35 @@ def _run_stmt_on_isabelle_server(self, stmt: str) -> None:
 
         # In proof mode. Execute tactics
         if self._proof_running or begin_clause:
-            stmt = stmt.strip()
+            stmt = f"{self.buffer}{stmt.strip()}"
 
-            # Run statement, with timeout. 
+            # Throw an error if "sorry" is used, which is not allowed
+            # Note that this is a fairly simple/optimistic way of handling it
+            if proof_search_mode and "sorry" in stmt:
+                raise Exception('Error: expected tactic, got "sorry". Do not use "sorry" in your proof.')
+ 
             try:
-                description = self.pisa_env.step(start_state, stmt, end_state, delete_old_state=False, forceTimeout=self.timeout_in_sec)
-            except FunctionTimedOut:
-                raise Exception("Error: the tactic timed out.")
-            if description.startswith('Step error:'):
-                raise Exception(description)
-            
+                # Run statement. TODO: pass in timeout
```

Comment:
I had to remove timeouts from the client side because they don’t work properly. Unfortunately, it seems pretty messy to pass in the desired timeout to PISA because it requires some complicated Python ↔ Scala interfacing. For now, I'm just changing the timeout parameter manually on line 587 in PisaOS.scala.